flowmachine/flowmachine/features/subscriber/topup_balance.py

```diff
@@ -0,0 +1,234 @@
+                FROM ({self.unioned_query.get_query()}) AS U
+                GROUP BY subscriber
+                UNION ALL
+                SELECT subscriber, {self.statistic}(post_event_balance) AS balance
+                FROM ({self.unioned_query.get_query()}) AS U
+                GROUP BY subscriber
+            ) U
+            GROUP BY subscriber
+            """
+            return sql
+
+        weighted_sum = f"SUM(weight * balance)"
+        weighted_avg = f"{weighted_sum} / SUM(weight)"
+        weighted_var = f"(SUM(weight * (balance ^ 2)) - ((({weighted_sum}) ^ 2) / SUM(weight))) / (SUM(weight) - 1)"
+        weighted_stddev = f"SQRT({weighted_var})"
+
+        statistic_clause = ""
+        if self.statistic in {"sum"}:
+            statistic_clause = weighted_sum
+        elif self.statistic in {"avg"}:
+            statistic_clause = weighted_avg
+        elif self.statistic in {"stddev"}:
+            statistic_clause = weighted_stddev
+        elif self.statistic in {"variance"}:
+            statistic_clause = weighted_var
+
+        # We calculate simple arithmetic average balances between events using
+        # the pre-event balance of the current observation and the post-event
+        # balance of the previous observation. Their respective weights are the
+        # number of seconds between both events. We calculate those variables
+        # using window functions.  The lag function will return NULL for the
+        # first event of each series.  In that case we assume that the
+        # pre-event balance equals to the pre-event balance of this first
+        # event. The generate weights are the number of seconds between this
+        # event and the beginning of the observed period. To account for the
+        # period between the last event and the end of the observed period we
+        # generate an additional row in which the average balance is equal to
+        # the post-event balance of the last observed top-up event. The weights
+        # are the number of seconds between this event and the end of the
+        # observed period.
+        weight_extraction_query = f"""
+        WITH W AS (
+            SELECT
+                msisdn AS subscriber,
+                (
+                    pre_event_balance +
+                    (LAG(post_event_balance, 1, pre_event_balance) OVER (PARTITION BY msisdn ORDER BY datetime))
+                ) / 2 AS balance,
+                post_event_balance,
+                datetime,
+                EXTRACT(EPOCH FROM
+                    datetime -
+                    LAG(
+                        datetime,
+                        1,
+                        TIMESTAMPTZ '{self.start}'
+                    ) OVER (PARTITION BY msisdn ORDER BY datetime)
+                ) AS weight,
+                CUME_DIST() OVER (PARTITION BY msisdn ORDER BY datetime) AS cume_dist
+            FROM (select * from events.topups) AS U
```

Comment:
```suggestion
            FROM (select * from events.topups) AS U
            WINDOW msisdn_by_datetime AS (PARTITION BY msisdn ORDER BY datetime)
```

Will allow you to do `OVER msisdn_by_datetime` for the other partitions.